### PR TITLE
Initial structure of bus business logic

### PR DESCRIPTION
**Pull request by Uneatablee**

### CODE REVIEW

File: src/business_logic/bus_stop/include/IVehicle.hpp
Comment (Mat3oo) [inline src/business_logic/bus_stop/include/IVehicle.hpp:1]: There is no built in concept of interfaces in c++, but you can achieve behavior of interfaces with abstract class. 
https://accu.org/journals/overload/12/62/radford_233/

File: src/business_logic/bus_stop/include/IVehicle.hpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        int current_position;     //number of current bus_stop
+        int direction;
+
+    public:
+
+        IVehicle(int vlenght, std::string v_name) : length(vlenght), name(v_name)
+        {std::cout << std::endl << "Creating Vehicle..";}
+        virtual ~IVehicle(){};
+        virtual void ImoveVehicle(int direction) = 0;
+        virtual void IdrawVehicle() = 0;
```
Comment (Mat3oo) [inline src/business_logic/bus_stop/include/IVehicle.hpp:23]: Interface method name should omit 'I' prefix.

**Review by Mat3oo — CHANGES_REQUESTED** [review]

File: src/business_logic/bus_stop/include/Bus_stop_extended.hpp [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#pragma once
+#include "IBus_stop.hpp"
+#include <memory>
+
+namespace cpp_playground::pg_business_logic::bus_stop
+{
+    class Bus_stop_extended : virtual public IBus_stop
```
Comment (Mat3oo) [inline src/business_logic/bus_stop/include/Bus_stop_extended.hpp:7]: If this is base class, name should be `Bus_stop_base` to not be misleading.

File: src/business_logic/bus_stop/include/Route.hpp
Comment (Mat3oo) [inline src/business_logic/bus_stop/include/Route.hpp:1]: Be consistent with your naming convention. Use _CamelCase_ or _pascalCase_, don't mix them. Most important is to pick one and stick with it :)

File: src/business_logic/bus_stop/include/Route.hpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+{
+    class Route final
+    {
+    private:
+
+        std::vector<std::unique_ptr<IBus_stop>> stops_list;
+
+    public:
+        Route()
+        {
```
Comment (Mat3oo) [inline src/business_logic/bus_stop/include/Route.hpp:16]: Move implementation to cpp file.

File: src/business_logic/bus_stop/src/Tbus.cpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
+#include "../include/Tbus.hpp"
+
+namespace cpp_playground::pg_business_logic::bus_stop
+{
+    void Tbus::moveVehicle(Vehicle_extended::Movement direction)
```
Comment (Mat3oo) [inline src/business_logic/bus_stop/src/Tbus.cpp:5]: I can see that this method is common for `Tbus` and `Tcar`, consider to move it to the base class.

File: src/business_logic/bus_stop/src/Tbus.cpp [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        }
+
+    }
+
+    void Tbus::drawVehicle()
+    {
+        std::cout << std::endl <<  "Drawing Vehicle..." << std::endl;
+    }
+
+    Tbus::~Tbus()
```
Comment (Mat3oo) [inline src/business_logic/bus_stop/src/Tbus.cpp:25]: Be compliant with rule of 0 or rule of 5. Define rest of compiler generated special methods.
(ref: https://en.cppreference.com/w/cpp/language/rule_of_three)

File: src/business_logic/bus_stop/src/Tcar.cpp [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        }
+
+    }
+
+    void Tcar::drawVehicle()
+    {
+        std::cout << std::endl <<  "Drawing Vehicle..." << std::endl;
+    }
+
+    Tcar::~Tcar()
```
Comment (Mat3oo) [inline src/business_logic/bus_stop/src/Tcar.cpp:25]: Be compliant with rule of 0 or rule of 5. Define rest of compiler generated special methods.
(ref: https://en.cppreference.com/w/cpp/language/rule_of_three)

File: src/business_logic/bus_stop/src/Tstop.cpp [hunk 7]
```diff
@@ -0,0 +1,27 @@
+#include "../include/Tstop.hpp"
+
+namespace cpp_playground::pg_business_logic::bus_stop
+{
+    Tstop::~Tstop(){}
```
Comment (Mat3oo) [inline src/business_logic/bus_stop/src/Tstop.cpp:5]: Be compliant with rule of 0 or rule of 5. If destructor doing nothing it should not even be defined.
(ref: https://en.cppreference.com/w/cpp/language/rule_of_three)

File: src/business_logic/bus_stop/src/Tterminal.cpp [hunk 8]
```diff
@@ -0,0 +1,27 @@
+#include "../include/Tterminal.hpp"
+
+namespace cpp_playground::pg_business_logic::bus_stop
+{
+    Tterminal::~Tterminal(){}
```
Comment (Mat3oo) [inline src/business_logic/bus_stop/src/Tterminal.cpp:5]: Be compliant with rule of 0 or rule of 5. If destructor doing nothing it should not even be defined.
(ref: https://en.cppreference.com/w/cpp/language/rule_of_three)

File: src/console/header.hpp
Comment (Mat3oo) [inline src/console/header.hpp:1]: There is a _precompiled header_ concept in c++ ;)
https://learn.microsoft.com/en-us/cpp/build/creating-precompiled-header-files?view=msvc-170

**Review by Mat3oo — CHANGES_REQUESTED** [review]

File: src/business_logic/bus_stop/include/Route.hpp [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+            stops_list.push_back(std::make_unique<Tterminal>("Bus terminal - 0"));
+            stops_list.push_back(std::make_unique<Tstop>("Bus stop - 1"));
+            stops_list.push_back(std::make_unique<Tstop>("Bus stop - 2"));
+            stops_list.push_back(std::make_unique<Tterminal>("Bus terminal - 3"));
+        }
+        ~Route()
```
[the remark this answers is not in the document]
Comment (Uneatablee) [reply src/business_logic/bus_stop/include/Route.hpp:22]: Changes applied

File: src/business_logic/bus_stop/include/Bus_stop_extended.hpp (hunk 2 above)
Comment (Uneatablee) [reply src/business_logic/bus_stop/include/Bus_stop_extended.hpp:7]: Changes applied

File: src/business_logic/bus_stop/include/Route.hpp
Comment (Uneatablee) [reply src/business_logic/bus_stop/include/Route.hpp:1]: Changes applied

File: src/business_logic/bus_stop/include/Vehicle_extended.hpp
[the remark this answers is not in the document]
Comment (Uneatablee) [reply src/business_logic/bus_stop/include/Vehicle_extended.hpp:1]: Changes Applied

File: src/business_logic/bus_stop/src/Tbus.cpp (hunk 5 above)
Comment (Uneatablee) [reply src/business_logic/bus_stop/src/Tbus.cpp:25]: rule of 0 -> changes applied

File: src/business_logic/bus_stop/src/Tcar.cpp (hunk 6 above)
Comment (Uneatablee) [reply src/business_logic/bus_stop/src/Tcar.cpp:25]: rule of 0 changes applied

File: src/business_logic/bus_stop/src/Tterminal.cpp (hunk 8 above)
Comment (Uneatablee) [reply src/business_logic/bus_stop/src/Tterminal.cpp:5]: changs applied

File: src/business_logic/bus_stop/src/Tstop.cpp (hunk 7 above)
Comment (Uneatablee) [reply src/business_logic/bus_stop/src/Tstop.cpp:5]: changes applied

File: src/business_logic/bus_stop/src/Tbus.cpp (hunk 4 above)
Comment (Uneatablee) [reply src/business_logic/bus_stop/src/Tbus.cpp:5]: changes applied

File: src/console/header.hpp
Comment (Uneatablee) [reply src/console/header.hpp:1]: added to learning plans

**Review by Mat3oo — APPROVED** [review]

**Pull request merged**